File: cpp/n299_GuessNum.cpp
// https://leetcode-cn.com/problems/bulls-and-cows/
#include <iostream>
#include <vector>
#include <sstream>
#include <string>
#include <unordered_map>
using namespace std;

class Solution {
public:
    string getHint(string secret, string guess) {
        int num_a = 0;
        int num_b = 0;
        unordered_map<char, int> word;
        for (size_t i = 0; i < secret.size(); ++i){
            if (secret[i] == guess[i]) num_a += 1;
            else if (!word.count(secret[i])) word[secret[i]] = 1;
            else word[secret[i]] += 1;
        }
        for (size_t i = 0; i < secret.size(); ++i){
            if (secret[i] == guess[i]) continue;
            else if (word.count(guess[i]) && word[guess[i]] >= 1){
                num_b += 1;
                word[guess[i]] -= 1;
            }
        }
        string res = to_string(num_a) + "A" + to_string(num_b) + "B";
        return res;
    }
};

int main(){
    string secret = "1807";
    string guess = "7810";
    Solution solu;
    cout << solu.getHint(secret, guess) << endl;
}

File: python/n299_GuessNum.py
# https://leetcode-cn.com/problems/bulls-and-cows/
class Solution:
    def getHint(self, secret: str, guess: str) -> str:
        num_a = num_b = 0
        word_dict = {}
        for i in range(len(secret)):
            if secret[i] == guess[i]:
                num_a += 1
            elif secret[i] not in word_dict.keys():
                word_dict[secret[i]] = 1
            else:
                word_dict[secret[i]] += 1
        for i in range(len(secret)):
            if secret[i] == guess[i]:
                continue
            else:
                if guess[i] in word_dict.keys() and word_dict[guess[i]] >= 1:
                    num_b += 1
                    word_dict[guess[i]] -= 1
                    
        res = str(num_a) + "A" + str(num_b) + "B"
        return res
        
        
if __name__ == "__main__":
    solu = Solution()
    secret = "19487290"
    guess = "29065513"
    print(solu.getHint(secret, guess))
                